### PR TITLE
Add multi-segment support for ELF loading in verilator_memutil

### PR DESCRIPTION
**Pull request by rswarbrick**
This is quite a long patch series because I wanted to make incremental changes (which should each work!), rather than a giant "rewrite the world" commit.

The point of the series is to allow us to run something like `Vmy_design --load-elf my_elf.elf` where `my_elf.elf` might have several segments, which might themselves target different memories in the design.

We make this possible by:
- Adding an optional memory location to each memory area that gets registered in the simulation. (in commit "*[memutil] Allow memory locations to have associated LMAs*")
- Adding a `--load-elf` option, which causes us to iterate over the segments, trying to find a matching memory based on LMA, and load everything up.

This patch has quite a few other commits, doing various bits of tidy-up or control inversion to make things fit better.

### CODE REVIEW

**Comment by rswarbrick** [comment]
Force-push rebases and (hopefully) fixes clang-format issue. Presumably, I'm locally running a different version from CI.

**Comment by rswarbrick** [comment]
Force-push fixes a path in the OTBN smoke test (missing `.elf` suffix) and fixes an assertion (we asserted that the memory size was greater than the data we loaded, but didn't account for the case where we don't know the memory size).

**Comment by rswarbrick** [comment]
1st force-push adds a missing include to sv_scoped.h. Second rebases.

File: hw/dv/verilator/cpp/verilator_memutil.cc
Comment (towoe) [reply hw/dv/verilator/cpp/verilator_memutil.cc:114]: In order to conform with the current structure, add a short option in
`getopt_long(argc, argv, ":r:m:f:l:E:h", long_options, nullptr);` or maybe with a small letter.

**Review by towoe — COMMENTED** [review]
Thanks a lot @rswarbrick for this improvement.
I still want to try some loading but here are already some nits and questions.

File: hw/dv/verilator/cpp/verilator_memutil.cc
Comment (towoe) [reply hw/dv/verilator/cpp/verilator_memutil.cc:114]: Maybe we do not want to do that.
I had some problems again with using short arguments multiple times, same problem happens with the long arguments without '='. I think the first `getopt_long()` invocation reorders `argv` which invalidates the options for an argument. Keeping it space free will keep the parts together and valid. I can do a follow up PR where we can discuss if we want to keep short arguments or remove then were necessary.

File: hw/dv/verilator/cpp/verilator_memutil.cc
Comment (rswarbrick) [reply hw/dv/verilator/cpp/verilator_memutil.cc:571]: It does, but I'm not sure that's a problem. In this function, I'm checking that `FindRegionForAddr` hasn't gone wrong. In the other check, I'm asserting that the `addr_to_mem_` map hasn't got corrupted somehow.

File: hw/dv/verilator/cpp/verilator_memutil.cc
Comment (rswarbrick) [reply hw/dv/verilator/cpp/verilator_memutil.cc:654]: The idea is that the `upper_bound` call finds the first region that starts strictly above `lma`. Then `std::prev(ub_it)` moves one to the left, so we should have the last region that starts at or before `lma`. The assertion is a data integrity check: we store the base of the memory area twice (once as a key in `addr_to_mem_` and once deep inside the value). It also acts as documentation: I've already edited this reply after realising the inequalities weren't coming out how I expected :-)

Anyway, the region we find starts at or before `lma`. But where does it finish? If the top of the region is also before `lma` then the load address has landed in a gap between regions (between `*m` and `ub_it->second`). In which case, we return null because there is no region that contains this address. If the top of the region is greater than `lma`, we've found a region with `base <= lma < top`. Success!

**Comment by rswarbrick** [comment]
First force-push changes things, responding to @towoe's comments. Second force-push rebases onto master (I split into two, hoping that the first diff might be helpful)

File: hw/dv/verilator/cpp/verilator_memutil.cc
Comment (rswarbrick) [reply hw/dv/verilator/cpp/verilator_memutil.cc]: Mid-air collision! Yep, sounds like a good idea. I've just changed things to match the other arguments, so at least it's uniform. I'm happy to lose the short forms here too.

File: hw/top_earlgrey/top_earlgrey_verilator.cc
Comment (towoe) [reply hw/top_earlgrey/top_earlgrey_verilator.cc:31]: Yes, lets leave it explicit.
I agree, adding the LMA to all memories in the future makes sense.

File: hw/dv/verilator/cpp/verilator_memutil.cc
Comment (towoe) [reply hw/dv/verilator/cpp/verilator_memutil.cc:251]: When I first started to code something together I think I looked at some online BSD manuals. Documentation seemed not to be that great.

Glad I could help :)

File: hw/dv/verilator/cpp/verilator_memutil.cc
Comment (towoe) [reply hw/dv/verilator/cpp/verilator_memutil.cc]: I think removing all short options would be helpful because we silently ignore errors from unknown short options due to the fact how `VerilatorSimCtrl::ParseCommandArgs` forwards the arguments to different parsers.
But I think it makes sense to do that in a separate PR in order to raise some more awareness for the change.
So for now you can either leave `-E` in here or it that again. Sorry for the confusion.

File: hw/dv/verilator/cpp/verilator_memutil.h
Comment (towoe) [reply hw/dv/verilator/cpp/verilator_memutil.h]: Sounds very reasonable :) Thanks

File: hw/dv/verilator/cpp/verilator_memutil.cc
Comment (towoe) [reply hw/dv/verilator/cpp/verilator_memutil.cc:662]: Shouldn't be a problem, was just curious.

**Comment by towoe** [comment]
As ELF loading depends on the addresses specified in the ELF file and multiple ELFs can be specified to load to different regions I think it might be helpful to print some info output for each section that is loaded into memory.
The reason is that a successive ELF load could overwrite a memory region which is already initialized and it is not really obvious what is happening.
Before this PR the memory written to was clearly specified by the arguments, with the changes it is somewhat hidden.
What do you think @rswarbrick ?

**Comment by rswarbrick** [comment]
Good idea about verbosity. I'd rather not print something unconditionally, so I've pushed up a version that adds a `--meminit-verbose` argument. Does that name sound reasonable?

The new code in the force-push is:
- "Use override keyword" (I noticed this and cleaned it up as I went past)
- "Parse all argument before loading anything" (so that "--raminit=foo --meminit-verbose" can print a message when loading `foo`)
- "Add a "verbose" flag to detail memory loads" (adding the `--meminit-verbose` flag)

And, of course, I re-plumbed the ELF loading code to support the verbose flag. Now, you can run:
```
build/lowrisc_ip_otbn_top_sim_0.1/sim-verilator/Votbn_top_sim --load-elf=build-bin/otbn/smoke_test/smoke.elf --meminit-verbose
Loading segment 0 from ELF file `build-bin/otbn/smoke_test/smoke.elf' into memory `imem'.
Loading segment 1 from ELF file `build-bin/otbn/smoke_test/smoke.elf' into memory `dmem'.
... (much) more output follows! ...
```
For symmetry, you also get messages with the existing `--meminit/--raminit/--fooinit` arguments.

**Comment by towoe** [comment]
> Good idea about verbosity. I'd rather not print something unconditionally, so I've pushed up a version that adds a `--meminit-verbose` argument. Does that name sound reasonable?
> 
Thanks a lot for addressing it.
I would be more in favor of having some default output for `--load-elf` as per my previous comment, but I leave that up to you. (I would argue that we currently already have some debug output and it would not add much more to it.)
I think `--meminit-verbose` could be misunderstood as people link that to `--meminit` and would disregard it for ELF loading.
But coming up with an alternative is not that easy :) `--verbose-init` `--verbose-startup`?
Or maybe it is enough to make it more explicit in the usage text that it applies to meminit and load-elf?

File: hw/dv/verilator/cpp/verilator_memutil.cc
Comment (towoe) [inline hw/dv/verilator/cpp/verilator_memutil.cc:662]: For consistency add the message from https://github.com/lowRISC/opentitan/pull/3544/files#diff-bf3d8ae05c262c011ba08283c82ea591R239 ?

**Review by towoe — APPROVED** [review]
Pending my comments which are optional this looks good to me.
I tried to run the otbn test which fails, but I think my env is not configured correctly.
Thanks @rswarbrick!

**Comment by rswarbrick** [comment]
Thanks for the review, @towoe.

> I would be more in favor of having some default output for --load-elf as per my previous comment, but I leave that up to you. (I would argue that we currently already have some debug output and it would not add much more to it.)

Well, there's not currently any debug output with the `--meminit`-based code unless you happen to have an empty or non-loadable segment. I think I probably prefer quiet by default.

> I think --meminit-verbose could be misunderstood as people link that to --meminit and would disregard it for ELF loading.
But coming up with an alternative is not that easy :) --verbose-init --verbose-startup?
Or maybe it is enough to make it more explicit in the usage text that it applies to meminit and load-elf?

The problem is that the meminit code hooks into the simulation's command line handling, which might support other arguments. This means that "verbose" needs something attached to make it clear that we're talking about the meminit code. I feel like `--verbose-init` and `--verbose-startup` are both generic: there are presumably other parts of the "startup" process for the simulation that might want to be verbose or not. How about `--verbose-mem-load`?

File: hw/dv/verilator/cpp/verilator_memutil.cc
Comment (rswarbrick) [reply hw/dv/verilator/cpp/verilator_memutil.cc:662]: I decided not to on purpose, because "load an ELF" has pretty well-defined semantics and you wouldn't ever expect to load a debug section or similar. Do you think there's a situation where this might be confusing?

Comment (towoe) [reply hw/dv/verilator/cpp/verilator_memutil.cc:662]: True. Did not consider this. It was just confusing to me with switching from meminit to load-elf because the message wasn't there. So I checked what was happening. But going forward this should not be a problem and you make a good point about it being not something anyone would ever expect.

Comment (towoe) [reply hw/dv/verilator/cpp/verilator_memutil.cc:662]: Maybe we should update the documentation to explain how the memory can be initialized and how this is happening.

**Comment by towoe** [comment]
> How about `--verbose-mem-load`?

Sounds good to me.

File: hw/dv/verilator/cpp/verilator_memutil.cc
Comment (rswarbrick) [reply hw/dv/verilator/cpp/verilator_memutil.cc:662]: Is this documented somewhere? Sorry, I didn't realise, otherwise I would have updated already.

Comment (towoe) [reply hw/dv/verilator/cpp/verilator_memutil.cc:662]: We have something in [docs.opentitan](https://docs.opentitan.org/doc/ug/getting_started_verilator/) but nothing specific.
Might be good to have some more details later, explaining what is happening for the different loading options.

Comment (rswarbrick) [reply hw/dv/verilator/cpp/verilator_memutil.cc:662]: Ah, right. I've just had a bit of a grep and it looks like this is really just in `getting_started_verilator.md`. We can't change that until we add an address map to the top-level (so there are LMAs for the memory sections). Let's leave it for now, I guess.

**Pull request merged**
